rust/gui-client/docs/process_split.md

```diff
@@ -1,34 +1,86 @@
+1. `/usr/lib/systemd/system/firezone-client.service` - The systemd service unit to auto-start the privileged client tunnel. Installed from the deb.
 
-## Binaries
+The binaries are separated because:
 
-I gave them new names to clarify.
+- If `firezone-client-tunnel` does not run as root, it may need `CAP_NET_ADMIN` set on the binary itself using `setcap`. We must not set that on the GUI binary.
+- Keeping the GUI code out of `firezone-client-tunnel` reduces the amount of code mapped into executable address space in the privileged tunnel process. It may not make a big difference, but all other things being equal, a smaller amount of code should be easier to secure.
 
-1. `firezoned` - A daemon that runs the tunnel and has elevated privilege
-2. `firezonectl` - A CLI program that runs unprivileged and exits once each operation is complete, like `systemctl` or `nmcli`
-3. `firezone-gui` - An unprivileged GUI program that keeps an icon in the system tray / notification center and is similar to `firezonectl`
+Other than those reasons, the GUI and privileged tunnel could spawn from the same binary.
+
+## Servers
+
+`/usr/lib/systemd/system/firezone-client.service` starts `firezone-client-tunnel --no-ipc`. Only service account tokens can be used. The only interaction is through SIGHUP or restarting the tunnel.
 
 ## Docker
 
-`firezoned` runs all the time. You can do `docker exec` to run `firezonectl` within the container.
+Same as a server, but Docker runs the process directly instead of using systemd.
+
+## Interactive CLI
+
+Same as a server, but a human runs the process instead of systemd.
 
 ## Desktop
 
-`firezoned` is started as a systemd service. It tries to find a service account token and use that. If there is no service account token, it waits for a GUI to connect over IPC.
+Same as a server, but without the `--no-ipc` flag. `firezone-gui` attempts to connect to the tunnel process. If there is a service account token, the IPC connection fails, and `firezone-gui` shows an error. If the IPC connection succeeds, `firezone-gui` can perform interactive auth or reload the token from the secure keyring to share it with `firezone-client-tunnel`.
+
+`firezone-gui --register-auto-start` puts a link in `$HOME/.config/autostart` to auto-start the program when the user logs in to their desktop.
+
+## Tokens
+
+`firezone-client-tunnel` tries to get a token in this order:
+
+- (The token must not be sent through CLI args)
+- If the `FIREZONE_TOKEN` env var is set, we copy that to memory and call `std::env::remove_var` so it doesn't stay in the environment longer than it needs to. (Per <https://security.stackexchange.com/questions/197784/is-it-unsafe-to-use-environmental-variables-for-secret-data>)
+- If there's a (service account) token in `/etc/dev.firezone.client/token`, and it's only readable by root, it uses that. (If anyone else can read it, it throws an error asking the user to fix the permissions and regenerate the token)
```

Comment:
```suggestion
- If there's a (service account) token in `/etc/firezone/token`, and it's only readable by root, it uses that. (If anyone else can read it, it throws an error asking the user to fix the permissions and regenerate the token)
```

I think this is more linux-y.

Reply:
Call it `client_token` if you want to reserve the gateway being able to read it from the same dir.